staging/src/k8s.io/apiserver/pkg/storageversion/updater.go

```diff
@@ -35,23 +35,96 @@ type Client interface {
+// condition based on the StorageVersions.
+func SetCommonEncodingVersion(sv *v1alpha1.StorageVersion) {
+	var oldCommonEncodingVersion *string
+	if sv.Status.CommonEncodingVersion != nil {
+		version := *sv.Status.CommonEncodingVersion
+		oldCommonEncodingVersion = &version
+
+	}
 	if len(sv.Status.StorageVersions) == 0 {
+		sv.Status.CommonEncodingVersion = nil
+	} else {
+		firstVersion := sv.Status.StorageVersions[0].EncodingVersion
+		agreed := true
+		for _, ssv := range sv.Status.StorageVersions {
+			if ssv.EncodingVersion != firstVersion {
+				agreed = false
+				break
+			}
+		}
+		if agreed {
+			sv.Status.CommonEncodingVersion = &firstVersion
+		} else {
+			sv.Status.CommonEncodingVersion = nil
+		}
+	}
+
+	condition := v1alpha1.StorageVersionCondition{
+		Type:               v1alpha1.AllEncodingVersionsEqual,
+		Status:             v1alpha1.ConditionFalse,
+		ObservedGeneration: sv.Generation,
+		LastTransitionTime: metav1.NewTime(time.Now()),
+		Reason:             "CommonEncodingVersionUnset",
+		Message:            "Common encoding version unset",
+	}
+	if sv.Status.CommonEncodingVersion != nil {
+		condition.Status = v1alpha1.ConditionTrue
+		condition.Reason = "CommonEncodingVersionSet"
+		condition.Message = "Common encoding version set"
+	}
+	forceTransition := false
+	if oldCommonEncodingVersion != nil && sv.Status.CommonEncodingVersion != nil &&
+		*oldCommonEncodingVersion != *sv.Status.CommonEncodingVersion {
+		forceTransition = true
+	}
+	setStatusCondition(&sv.Status.Conditions, condition, forceTransition)
+}
+
+func findStatusCondition(conditions []v1alpha1.StorageVersionCondition,
+	conditionType v1alpha1.StorageVersionConditionType) *v1alpha1.StorageVersionCondition {
+	for i := range conditions {
+		if conditions[i].Type == conditionType {
+			return &conditions[i]
+		}
+	}
+	return nil
+}
+
+// setStatusCondition sets the corresponding condition in conditions to newCondition.
+// conditions must be non-nil.
+// 1. if the condition of the specified type already exists (all fields of the existing condition are updated to
+//    newCondition, LastTransitionTime is set to now if the new status differs from the old status)
+// 2. if a condition of the specified type does not exist (LastTransitionTime is set to now() if unset, and newCondition is appended)
+// NOTE: forceTransition allows overwriting LastTransitionTime even when the status doesn't change.
+func setStatusCondition(conditions *[]v1alpha1.StorageVersionCondition, newCondition v1alpha1.StorageVersionCondition,
+	forceTransition bool) {
+	if conditions == nil {
 		return
 	}
-	firstVersion := sv.Status.StorageVersions[0].EncodingVersion
-	agreed := true
-	for _, ssv := range sv.Status.StorageVersions {
-		if ssv.EncodingVersion != firstVersion {
-			agreed = false
-			break
+	existingCondition := findStatusCondition(*conditions, newCondition.Type)
+	if existingCondition == nil {
+		if newCondition.LastTransitionTime.IsZero() {
```

Comment:
the function is copied from the [condition helper](https://github.com/kubernetes/kubernetes/blob/c678434623be4957d892a9865e5649f887a40c49/staging/src/k8s.io/apimachinery/pkg/api/meta/conditions.go#L25-L55) (we need our own copy due to type safety). Let's keep this branch because this function can be used for other storage version conditions in future